### PR TITLE
Profile version

### PR DESCRIPTION
**Pull request by TheMarex**
# Issue

Add explicit versioning to the profiles. Addresses #3384 

This is currently unused but could be utilized to activate other API features only when explicitly bumping the version. This will be used to make #2399 backwards compatible.

## Tasklist
 - [x] review
 - [x] adjust for comments

## Requirements / Relations

Requirement for #2399

### CODE REVIEW

File: profiles/car.lua [hunk 1]
```diff
@@ -194,6 +194,7 @@ properties.max_speed_for_map_matching      = 180/3.6 -- 180kmph -> m/s
 properties.use_turn_restrictions           = true
 properties.continue_straight_at_waypoint   = true
 properties.left_hand_driving               = false
+properties.version                         = 0
```
Comment (daniel-j-h) [inline profiles/car.lua]: Hrm should we call it `profile_api_version` or something like that? Not sure if version alone is telling for users.

Comment (TheMarex) [inline profiles/car.lua]: Actually now that you mention it, maybe we should not even put it in `ProfileProperties`.

Do you think a global variable like `profile_api_version = 0` would be too obscure?

Comment (TheMarex) [inline profiles/car.lua]: My thinking here is that we actually don't need the profile version information outside of the scripting context.

Comment (daniel-j-h) [inline profiles/car.lua]: Hm maybe at the top of the profile, like CMake does it :+1:

**Comment by oxidase** [comment]
@TheMarex maybe not related to the PR, how to do testing in future? Should tests be replicated for every profile version (actually bad idea), just maintain the actual profile version (no tests for deprecated profile versions) or some tests added to check only profile functions?

**Comment by TheMarex** [comment]
@oxidase that is a good question. I the most feasible solution would be to only have full tests for the current version, but specific tests for functionality changes between versions.

For example in the context of the #2399 PR we should have all our tests run on `api_version = 1` (e.g. the new version). But should have a test that `api_version = 0`-style `turn_function` and `segment_function` still work.

**Comment by TheMarex** [comment]
Moved it to a global variable and enforced correct version over a check.

**Comment by oxidase** [comment]
@TheMarex just [a simple test](https://gist.github.com/oxidase/cfe6287d069007415905b73397fda86c) to make codecov green

**Comment by TheMarex** [comment]
@oxidase the test looks good. 👍 you can just push to the branch, that's fine.

**Comment by oxidase** [comment]
@TheMarex pushed. `Not-defined API version` test expects that osrm-extract is not failing if api-version is not defined. api-version 0 will be used as default initialized.

**Comment by emiltin** [comment]
should version be in the format 1.0.0?

**Comment by TheMarex** [comment]
@emiltin I was think using the same scheme as in the HTTP API, which also uses simple increments. Minor/patch versions are not that useful for single components because you typically only care about breaking changes.

File: src/extractor/scripting_environment_lua.cpp [hunk 2]
```diff
@@ -306,6 +306,20 @@ void Sol2ScriptingEnvironment::InitContext(LuaScriptingContext &context)
     context.has_node_function = node_function.valid();
     context.has_way_function = way_function.valid();
     context.has_segment_function = segment_function.valid();
+    auto maybe_version = context.state.get<sol::optional<int>>("api_version");
+    if (maybe_version)
+     {
```
Comment (daniel-j-h) [inline src/extractor/scripting_environment_lua.cpp]: Looks like the formatting is off here ..

**Pull request merged**